src/typeInference/annotateAST.ts

```diff
@@ -0,0 +1,158 @@
+import { SequenceMultiplicity, SequenceType, ValueType } from '../expressions/dataTypes/Value';
+import StaticContext from '../expressions/StaticContext';
+import astHelper, { IAST } from '../parsing/astHelper';
+import { annotateBinOp } from './annotateBinaryOperator';
+import { annotateCastableOperator, annotateCastOperator } from './annotateCastOperators';
+import {
+	annotateGeneralCompare,
+	annotateNodeCompare,
+	annotateValueCompare,
+} from './annotateCompareOperator';
+import { annotateFunctionCall } from './annotateFunctionCall';
+import { annotateLogicalOperator } from './annotateLogicalOperator';
+import { annotateUnaryMinus, annotateUnaryPlus } from './annotateUnaryOperator';
+
+/**
+ * Recursively traverse the AST in the depth first, pre-order to infer type and annotate AST;
+ * Annotates as much type information as possible to the AST nodes.
+ * Inserts attribute `type` to the corresponding node if type is inferred.
+ *
+ * @param ast The AST to annotate
+ * @param staticContext The static context used for function lookups
+ */
+export default function annotateAst(ast: IAST, staticContext?: StaticContext) {
+	annotate(ast, staticContext);
+}
+
+/**
+ * Recursively traverse the AST in the depth first, pre-order to infer type and annotate AST;
+ * Annotates as much type information as possible to the AST nodes.
+ * Inserts attribute `type` to the corresponding node if type is inferred.
+ *
+ * @param ast The AST to annotate.
+ * @param staticContext The static context to use for function lookups.
+ * @throws errors when attempts to annotate fail.
+ * @returns The type of the AST node or `undefined` when the type cannot be annotated.
+ */
+export function annotate(ast: IAST, staticContext: StaticContext): SequenceType | undefined {
+	// Check if we actually have an AST
+	if (!ast) {
+		return undefined;
```

Comment:
how should this be hit? I assume the ast argument is just there, right?

Reply:
This function was sometimes called from the default case of annotate so I moved the check there.